openmls/src/treesync/node/leaf_node.rs

```diff
@@ -263,53 +285,122 @@ impl LeafNode {
         signature_keys: &[SignaturePublicKey],
         encryption_keys: &[EncryptionKey],
         members_supported_credentials: &[&[CredentialType]],
         currently_in_use: &[CredentialType],
     ) -> Result<&Self, LeafNodeValidationError> {
-        self.validate_required_capabilities(required_capabilities)?
-            .validate_that_capabilities_contain_extension_types()?
+        // The client verifies the validity of a LeafNode using the following steps:
+        self
+            // Verify that the credential in the LeafNode is valid as described in Section 5.3.1.
+            // ...
+            // Verify that the signature on the LeafNode is valid using signature_key.
+            // ...
+            // Verify that the LeafNode is compatible with the group's parameters.
+            // (ValSem106)
+            .validate_version_ciphersuite(version, ciphersuite)?
+            // If the GroupContext has a required_capabilities extension, then the required extensions, proposals, and credential types MUST be listed in the LeafNode's capabilities field.
+            // (ValSem106)
+            .validate_required_capabilities(required_capabilities)?
             .validate_that_capabilities_contain_credential_type()?
-            .validate_that_signature_key_is_unique(signature_keys)?
-            .validate_that_encryption_key_is_unique(encryption_keys)?
+            // Verify that the credential type is supported by all members of the group, as specified by the capabilities field of each member's LeafNode, ...
             .validate_against_group_credentials(members_supported_credentials)?
-            .validate_credential_in_use(currently_in_use)?;
+            // ... and that the capabilities field of this LeafNode indicates support for all the credential types currently in use by other members.
+            .validate_credential_in_use(currently_in_use)?
+            // Verify that the extensions in the LeafNode are supported by checking that the ID for each extension in the extensions field is listed in the capabilities.extensions field of the LeafNode.
+            .validate_that_capabilities_contain_extension_types()?
+            // Verify that the following fields are unique among the members of the group:
+            // signature_key (ValSem104)
+            .validate_that_signature_key_is_unique(signature_keys)?
+            // encryption_key (ValSem114)
+            .validate_that_encryption_key_is_unique(encryption_keys)?;
 
         Ok(self)
     }
 
+    /// ValSem106: Check the required capabilities of the add proposals
+    // ~ - Are the two listed in the `Capabilities` Extension?
+    pub(crate) fn validate_version_ciphersuite(
+        &self,
+        version: ProtocolVersion,
+        ciphersuite: Ciphersuite,
+    ) -> Result<&Self, LeafNodeValidationError> {
+        if !self.payload.capabilities.versions.contains(&version) {
+            Err(LeafNodeValidationError::UnsupportedVersion)
```

Comment:
If you use returns, the function becomes more readable because you don't need all the nesting.